rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,869 @@
+The data field is of type `RawExtension`, allowing it to store any policy type.
+
+The Policy Server replicas will be watching the new `PolicyRevision` CRD directly and it will change the internal state of the evaluation environment when a new policy is created or updated.
+
+#### PolicyServer leader reconciler
+
+A leader election mechanism will be implemented to ensure that only one Policy Server instance is running the leader reconciler at a time.
+The leader will be responsible to pull the policies from the registry, validate the policy settings, precompile the policy and store it in a shared cache. See [Shared Policy Cache](#shared-policy-cache) for more details.
+We will rely on [Kubernetes Leases](https://kubernetes.io/docs/concepts/architecture/leases/) for this functionality, using the same mechanism employed by controllers scaffolded with Kubebuilder to elect a leader.
+Since the Policy Server is implemented in Rust, we will use the [kubert](https://docs.rs/kubert/0.22.0/kubert/lease/index.html) crate's lease module for this purpose.
+These features are not provided by the `kube-rs` crate, but kubert is built on top of `kube-rs`, is actively maintained, and is used in production by Linkerd.
+
+After the policy is precompiled, the leader will change the status condition of type `Initialized` of the `PolicyRevision` to `True`.
+This will trigger the replica reconciler to load the policy in the evaluation environment of all the policy server pods that are part of the StatefulSet.
+
+#### Policy Server replica reconciler
+
+All the replicas, including the leader, will be running a reconciliation loop to configure the policies in the evaluation environment.
+The replicas will be watching the `PolicyRevision` CRDs and load the policy in the evaluation environment when the status condition of type `Initialized` is set to `True` and the `enabled` field is set to `true`.
+In case of an error, the policy status will be updated, but the Policy Server will continue serving the previous valid policy until the issue is resolved.
+This is possible because multiple `PolicyRevisions` can be enabled at the same time.
+The Policy Server provides access to current policies through the `/validate` endpoint, incorporating the `generation` number as a URL parameter.
+This setup enables the webhook to target a specific policy generation.
+For example, the endpoint `/validate/privileged-pods/2` validates requests using the `privileged-pods` policy at generation 2.
+
+#### Policy and PolicyRevision conditions
+
+The `PolicyRevision` CRD will have a status field to store the status of the policy.
+The status will be propagated to the `Policy` CRD by the controller, so that the user can monitor the status of the policy without having to access the `PolicyRevision` CRD.
+As the `PolicyRevision` CRD is a namespaced resource, only admin users will have the right permissions to access it.
```

Comment:
The Policy could have 2 new spec fields that point to the latest PolicyRevision, and the active PolicyRevision. This may simplify garbage collection, general implementation, and user UI on knowing exactly what Policy is being run (settings, etc).